Test Failure (20200424-0751): junit.framework.TestSuite.com.ibm.ws.microprofile.reactive.streams.test.suite.ReactiveJaxRSTest

```
componenttest.exception.TopologyException: Failures occured while waiting for app to start:
  Application ReactiveWithJaxRS failure: CWWKZ0002E: An exception occurred while starting the application ReactiveWithJaxRS. The exception message was: com.ibm.ws.container.service.state.StateChangeException: java.lang.NullPointerException
	at componenttest.topology.impl.LibertyServer.validateAppsLoaded(LibertyServer.java:1624)
	at componenttest.topology.impl.LibertyServer.validateServerStarted(LibertyServer.java:2153)
	at componenttest.topology.impl.LibertyServer.startServerWithArgs(LibertyServer.java:1437)
	at componenttest.topology.impl.LibertyServer.startServerAndValidate(LibertyServer.java:1060)
	at componenttest.topology.impl.LibertyServer.startServerAndValidate(LibertyServer.java:1042)
	at componenttest.topology.impl.LibertyServer.startServerAndValidate(LibertyServer.java:1029)
	at componenttest.topology.impl.LibertyServer.startServer(LibertyServer.java:877)
	at com.ibm.ws.microprofile.reactive.streams.test.suite.ReactiveJaxRSTest.setUp(ReactiveJaxRSTest.java:60)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:314)
	at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:167)
```

From the logs,
```
com.ibm.ws.container.service.state.StateChangeException: java.lang.NullPointerException
	at com.ibm.ws.container.service.state.internal.ApplicationStateManager.fireStarting(ApplicationStateManager.java:55)
	at com.ibm.ws.container.service.state.internal.StateChangeServiceImpl.fireApplicationStarting(StateChangeServiceImpl.java:50)
	at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.preDeployApp(SimpleDeployedAppInfoBase.java:547)
	at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.installApp(SimpleDeployedAppInfoBase.java:508)
	at com.ibm.ws.app.manager.module.internal.DeployedAppInfoBase.deployApp(DeployedAppInfoBase.java:347)
	at com.ibm.ws.app.manager.war.internal.WARApplicationHandlerImpl.install(WARApplicationHandlerImpl.java:65)
	at com.ibm.ws.app.manager.internal.statemachine.StartAction.execute(StartAction.java:144)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.enterState(ApplicationStateMachineImpl.java:1295)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.run(ApplicationStateMachineImpl.java:886)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:239)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NullPointerException
	at com.ibm.ws.context.service.serializable.ThreadContextDescriptorImpl.<init>(ThreadContextDescriptorImpl.java:197)
	at com.ibm.ws.context.service.serializable.ThreadContextManager.captureThreadContext(ThreadContextManager.java:123)
	at com.ibm.ws.concurrent.internal.ContextServiceImpl.captureThreadContext(ContextServiceImpl.java:281)
	at com.ibm.ws.concurrent.internal.ContextServiceImpl.captureThreadContext(ContextServiceImpl.java:224)
	at com.ibm.ws.concurrent.internal.ManagedExecutorServiceImpl.createCallbacks(ManagedExecutorServiceImpl.java:346)
	at com.ibm.ws.concurrent.internal.ManagedExecutorServiceImpl.invokeAll(ManagedExecutorServiceImpl.java:467)
	at org.jboss.weld.executor.AbstractExecutorServices.invokeAllAndCheckForExceptions(AbstractExecutorServices.java:59)
	at org.jboss.weld.executor.AbstractExecutorServices.invokeAllAndCheckForExceptions(AbstractExecutorServices.java:67)
	at org.jboss.weld.bootstrap.ConcurrentBeanDeployer.addClasses(ConcurrentBeanDeployer.java:52)
	at org.jboss.weld.bootstrap.BeanDeployment.createClasses(BeanDeployment.java:198)
	at org.jboss.weld.bootstrap.WeldStartup.startInitialization(WeldStartup.java:412)
	at org.jboss.weld.bootstrap.WeldBootstrap.startInitialization(WeldBootstrap.java:79)
	at com.ibm.ws.cdi.impl.CDIContainerImpl$2.run(CDIContainerImpl.java:149)
	at com.ibm.ws.cdi.impl.CDIContainerImpl$2.run(CDIContainerImpl.java:146)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:678)
	at com.ibm.ws.cdi.impl.CDIContainerImpl.startInitialization(CDIContainerImpl.java:146)
	at com.ibm.ws.cdi.liberty.CDIRuntimeImpl.applicationStarting(CDIRuntimeImpl.java:453)
	at com.ibm.ws.container.service.state.internal.ApplicationStateManager.fireStarting(ApplicationStateManager.java:51)
	... 12 more
```

Brent Daniel looked into this and believes that the issue is caused by the addition of the JavaEEVersion reference to ThreadContextManager as optional and static, interimttently causing a bounce when the JavaEEVersion arrives.

I'll look into whether it can be switched to dynamic.

--- a/dev/com.ibm.ws.concurrent.persistent/src/com/ibm/ws/concurrent/persistent/internal/PersistentExecutorImpl.java
+++ b/dev/com.ibm.ws.concurrent.persistent/src/com/ibm/ws/concurrent/persistent/internal/PersistentExecutorImpl.java
@@ -197,9 +197,14 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
     private final Map<String, String> defaultExecProps = new TreeMap<String, String>();
 
     /**
-     * Jakarta EE versiom if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
+     * Jakarta EE version if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
      */
-    private int eeVersion;
+    private volatile int eeVersion;
+
+    /**
+     * Tracks the most recently bound EE version service reference. Only use this within the set/unsetEEVersion methods.
+     */
+    private ServiceReference<JavaEEVersion> eeVersionRef;
 
     /**
      * Common Liberty thread pool.
@@ -1775,7 +1780,7 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
      */
     @Reference(service = JavaEEVersion.class,
                cardinality = ReferenceCardinality.OPTIONAL,
-               policy = ReferencePolicy.STATIC,
+               policy = ReferencePolicy.DYNAMIC,
                policyOption = ReferencePolicyOption.GREEDY)
     protected void setEEVersion(ServiceReference<JavaEEVersion> ref) {
         String version = (String) ref.getProperty("version");
@@ -1786,6 +1791,7 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
             String major = dot > 0 ? version.substring(0, dot) : version;
             eeVersion = Integer.parseInt(major);
         }
+        eeVersionRef = ref;
     }
 
     /**
@@ -2039,7 +2045,10 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
      * @param ref reference to the service
      */
     protected void unsetEEVersion(ServiceReference<JavaEEVersion> ref) {
-        eeVersion = 0;
+        if (eeVersionRef == ref) {
+            eeVersionRef = null;
+            eeVersion = 0;
+        }
     }
 
     /**

--- a/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ContextServiceImpl.java
+++ b/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ContextServiceImpl.java
@@ -118,9 +118,14 @@ public class ContextServiceImpl implements //
     private ComponentContext componentContext;
 
     /**
-     * Jakarta EE versiom if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
+     * Jakarta EE version if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
      */
-    private int eeVersion;
+    private volatile int eeVersion;
+
+    /**
+     * Tracks the most recently bound EE version service reference. Only use this within the set/unsetEEVersion methods.
+     */
+    private ServiceReference<JavaEEVersion> eeVersionRef;
 
     /**
      * Lock for reading and updating configuration.
@@ -762,7 +767,7 @@ public class ContextServiceImpl implements //
      */
     @Reference(service = JavaEEVersion.class,
                cardinality = ReferenceCardinality.OPTIONAL,
-               policy = ReferencePolicy.STATIC,
+               policy = ReferencePolicy.DYNAMIC,
                policyOption = ReferencePolicyOption.GREEDY)
     protected void setEEVersion(ServiceReference<JavaEEVersion> ref) {
         String version = (String) ref.getProperty("version");
@@ -773,6 +778,7 @@ public class ContextServiceImpl implements //
             String major = dot > 0 ? version.substring(0, dot) : version;
             eeVersion = Integer.parseInt(major);
         }
+        eeVersionRef = ref;
     }
 
     /**
@@ -809,7 +815,10 @@ public class ContextServiceImpl implements //
      * @param ref reference to the service
      */
     protected void unsetEEVersion(ServiceReference<JavaEEVersion> ref) {
-        eeVersion = 0;
+        if (eeVersionRef == ref) {
+            eeVersionRef = null;
+            eeVersion = 0;
+        }
     }
 
     /**

--- a/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ManagedExecutorServiceImpl.java
+++ b/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ManagedExecutorServiceImpl.java
@@ -135,9 +135,14 @@ public class ManagedExecutorServiceImpl implements ExecutorService, ManagedExecu
     private final AtomicReference<Map<String, String>> defaultExecutionProperties = new AtomicReference<Map<String, String>>();
 
     /**
-     * Jakarta EE versiom if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
+     * Jakarta EE version if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
      */
-    int eeVersion;
+    volatile int eeVersion;
+
+    /**
+     * Tracks the most recently bound EE version service reference. Only use this within the set/unsetEEVersion methods.
+     */
+    private ServiceReference<JavaEEVersion> eeVersionRef;
 
     /**
      * Hash code for this instance.
@@ -567,7 +572,7 @@ public class ManagedExecutorServiceImpl implements ExecutorService, ManagedExecu
      */
     @Reference(service = JavaEEVersion.class,
                cardinality = ReferenceCardinality.OPTIONAL,
-               policy = ReferencePolicy.STATIC,
+               policy = ReferencePolicy.DYNAMIC,
                policyOption = ReferencePolicyOption.GREEDY)
     protected void setEEVersion(ServiceReference<JavaEEVersion> ref) {
         String version = (String) ref.getProperty("version");
@@ -578,6 +583,7 @@ public class ManagedExecutorServiceImpl implements ExecutorService, ManagedExecu
             String major = dot > 0 ? version.substring(0, dot) : version;
             eeVersion = Integer.parseInt(major);
         }
+        eeVersionRef = ref;
     }
 
     /**
@@ -717,7 +723,10 @@ public class ManagedExecutorServiceImpl implements ExecutorService, ManagedExecu
      * @param ref reference to the service
      */
     protected void unsetEEVersion(ServiceReference<JavaEEVersion> ref) {
-        eeVersion = 0;
+        if (eeVersionRef == ref) {
+            eeVersionRef = null;
+            eeVersion = 0;
+        }
     }
 
     /**

--- a/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ManagedScheduledExecutorServiceImpl.java
+++ b/dev/com.ibm.ws.concurrent/src/com/ibm/ws/concurrent/internal/ManagedScheduledExecutorServiceImpl.java
@@ -247,7 +247,7 @@ public class ManagedScheduledExecutorServiceImpl extends ManagedExecutorServiceI
     @Override
     @Reference(service = JavaEEVersion.class,
                cardinality = ReferenceCardinality.OPTIONAL,
-               policy = ReferencePolicy.STATIC,
+               policy = ReferencePolicy.DYNAMIC,
                policyOption = ReferencePolicyOption.GREEDY)
     protected void setEEVersion(ServiceReference<JavaEEVersion> ref) {
         super.setEEVersion(ref);

--- a/dev/com.ibm.ws.context/src/com/ibm/ws/context/service/serializable/ThreadContextManager.java
+++ b/dev/com.ibm.ws.context/src/com/ibm/ws/context/service/serializable/ThreadContextManager.java
@@ -64,9 +64,14 @@ public class ThreadContextManager implements WSContextService {
     private final Set<String> alwaysEnabled = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
 
     /**
-     * Jakarta EE versiom if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
+     * Jakarta EE version if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
      */
-    int eeVersion;
+    volatile int eeVersion;
+
+    /**
+     * Tracks the most recently bound EE version service reference. Only use this within the set/unsetEEVersion methods.
+     */
+    private ServiceReference<JavaEEVersion> eeVersionRef;
 
     /**
      * The metadata identifier service.
@@ -230,7 +235,7 @@ public class ThreadContextManager implements WSContextService {
      */
     @Reference(service = JavaEEVersion.class,
                cardinality = ReferenceCardinality.OPTIONAL,
-               policy = ReferencePolicy.STATIC,
+               policy = ReferencePolicy.DYNAMIC,
                policyOption = ReferencePolicyOption.GREEDY)
     protected void setEEVersion(ServiceReference<JavaEEVersion> ref) {
         String version = (String) ref.getProperty("version");
@@ -241,6 +246,7 @@ public class ThreadContextManager implements WSContextService {
             String major = dot > 0 ? version.substring(0, dot) : version;
             eeVersion = Integer.parseInt(major);
         }
+        eeVersionRef = ref;
     }
 
     /**
@@ -276,7 +282,10 @@ public class ThreadContextManager implements WSContextService {
      * @param ref reference to the service
      */
     protected void unsetEEVersion(ServiceReference<JavaEEVersion> ref) {
-        eeVersion = 0;
+        if (eeVersionRef == ref) {
+            eeVersionRef = null;
+            eeVersion = 0;
+        }
     }
 
     /**

--- a/dev/com.ibm.ws.jca/bnd.bnd
+++ b/dev/com.ibm.ws.jca/bnd.bnd
@@ -127,7 +127,7 @@ Service-Component: \
     connectorService='com.ibm.ws.jca.cm.ConnectorService';\
     contextProvider='com.ibm.wsspi.threadcontext.jca.JCAContextProvider';\
     contextService='com.ibm.wsspi.threadcontext.WSContextService';\
-    EEVersion='com.ibm.ws.javaee.version.JavaEEVersion';\
+    eEVersion='com.ibm.ws.javaee.version.JavaEEVersion';\
     executorService='java.util.concurrent.ExecutorService';\
     futureMonitor='com.ibm.ws.threading.FutureMonitor';\
     requiredContextProvider='com.ibm.wsspi.threadcontext.jca.JCAContextProvider';\
@@ -136,9 +136,10 @@ Service-Component: \
     tranSyncRegistry='javax.transaction.TransactionSynchronizationRegistry';\
     jcaSecurityContextService=com.ibm.ws.jca.security.JCASecurityContext;\
     configuration-policy:=require;\
-    greedy:='beanValidationService,contextProvider,EEVersion,jcaSecurityContextService';\
+    dynamic:='eEVersion';\
+    greedy:='beanValidationService,contextProvider,eEVersion,jcaSecurityContextService';\
     multiple:='contextProvider,requiredContextProvider';\
-    optional:='beanValidationService,contextProvider,EEVersion,requiredContextProvider,jcaSecurityContextService';\
+    optional:='beanValidationService,contextProvider,eEVersion,requiredContextProvider,jcaSecurityContextService';\
     properties:="contextService.target=(id=unbound),executorService.target=(id=unbound),requiredContextProvider.target=(id=unbound),resourceAdapterService.target=(id=unbound)",\
   com.ibm.ws.jca.sslHelper;\
     implementation:=com.ibm.ws.jca.internal.SSLHelperImpl;\

--- a/dev/com.ibm.ws.jca/src/com/ibm/ws/jca/internal/BootstrapContextImpl.java
+++ b/dev/com.ibm.ws.jca/src/com/ibm/ws/jca/internal/BootstrapContextImpl.java
@@ -168,9 +168,14 @@ public class BootstrapContextImpl implements BootstrapContext, ApplicationRecycl
     private ServiceReference<WSContextService> contextSvcRef;
 
     /**
-     * Jakarta EE versiom if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
+     * Jakarta EE version if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
      */
-    int eeVersion;
+    volatile int eeVersion;
+
+    /**
+     * Tracks the most recently bound EE version service reference. Only use this within the set/unsetEEVersion methods.
+     */
+    private ServiceReference<JavaEEVersion> eeVersionRef;
 
     /**
      * Liberty executor
@@ -1024,6 +1029,7 @@ public class BootstrapContextImpl implements BootstrapContext, ApplicationRecycl
             String major = dot > 0 ? version.substring(0, dot) : version;
             eeVersion = Integer.parseInt(major);
         }
+        eeVersionRef = ref;
     }
 
     /**
@@ -1256,7 +1262,10 @@ public class BootstrapContextImpl implements BootstrapContext, ApplicationRecycl
      * @param ref reference to the service
      */
     protected void unsetEEVersion(ServiceReference<JavaEEVersion> ref) {
-        eeVersion = 0;
+        if (eeVersionRef == ref) {
+            eeVersionRef = null;
+            eeVersion = 0;
+        }
     }
 
     /**

--- a/dev/com.ibm.ws.microprofile.health.2.0/src/com/ibm/ws/microprofile/health20/services/impl/HealthCheck20ExecutorImpl.java
+++ b/dev/com.ibm.ws.microprofile.health.2.0/src/com/ibm/ws/microprofile/health20/services/impl/HealthCheck20ExecutorImpl.java
@@ -45,9 +45,14 @@ public class HealthCheck20ExecutorImpl implements HealthCheck20Executor {
     private AppModuleContextService appModuleContextService;
 
     /**
-     * Jakarta EE versiom if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
+     * Jakarta EE version if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
      */
-    private int eeVersion;
+    private volatile int eeVersion;
+
+    /**
+     * Tracks the most recently bound EE version service reference. Only use this within the set/unsetEEVersion methods.
+     */
+    private ServiceReference<JavaEEVersion> eeVersionRef;
 
     private HealthCheck20CDIBeanInvoker healthCheckCDIBeanInvoker;
     private J2EENameFactory j2eeNameFactory;
@@ -122,10 +127,7 @@ public class HealthCheck20ExecutorImpl implements HealthCheck20Executor {
      *
      * @param ref reference to the service
      */
-    @Reference(service = JavaEEVersion.class,
-               cardinality = ReferenceCardinality.OPTIONAL,
-               policy = ReferencePolicy.STATIC,
-               policyOption = ReferencePolicyOption.GREEDY)
+    @Reference(service = JavaEEVersion.class, cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
     protected void setEEVersion(ServiceReference<JavaEEVersion> ref) {
         String version = (String) ref.getProperty("version");
         if (version == null) {
@@ -135,6 +137,7 @@ public class HealthCheck20ExecutorImpl implements HealthCheck20Executor {
             String major = dot > 0 ? version.substring(0, dot) : version;
             eeVersion = Integer.parseInt(major);
         }
+        eeVersionRef = ref;
     }
 
     /**
@@ -143,6 +146,9 @@ public class HealthCheck20ExecutorImpl implements HealthCheck20Executor {
      * @param ref reference to the service
      */
     protected void unsetEEVersion(ServiceReference<JavaEEVersion> ref) {
-        eeVersion = 0;
+        if (eeVersionRef == ref) {
+            eeVersionRef = null;
+            eeVersion = 0;
+        }
     }
 }

--- a/dev/com.ibm.ws.microprofile.health/src/com/ibm/ws/microprofile/health/services/impl/HealthExecutorImpl.java
+++ b/dev/com.ibm.ws.microprofile.health/src/com/ibm/ws/microprofile/health/services/impl/HealthExecutorImpl.java
@@ -43,9 +43,14 @@ public class HealthExecutorImpl implements HealthExecutor {
     private AppModuleContextService appModuleContextService;
 
     /**
-     * Jakarta EE versiom if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
+     * Jakarta EE version if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
      */
-    private int eeVersion;
+    private volatile int eeVersion;
+
+    /**
+     * Tracks the most recently bound EE version service reference. Only use this within the set/unsetEEVersion methods.
+     */
+    private ServiceReference<JavaEEVersion> eeVersionRef;
 
     private HealthCheckCDIBeanInvoker healthCheckCDIBeanInvoker;
     private J2EENameFactory j2eeNameFactory;
@@ -120,10 +125,7 @@ public class HealthExecutorImpl implements HealthExecutor {
      *
      * @param ref reference to the service
      */
-    @Reference(service = JavaEEVersion.class,
-               cardinality = ReferenceCardinality.OPTIONAL,
-               policy = ReferencePolicy.STATIC,
-               policyOption = ReferencePolicyOption.GREEDY)
+    @Reference(service = JavaEEVersion.class, cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
     protected void setEEVersion(ServiceReference<JavaEEVersion> ref) {
         String version = (String) ref.getProperty("version");
         if (version == null) {
@@ -133,6 +135,7 @@ public class HealthExecutorImpl implements HealthExecutor {
             String major = dot > 0 ? version.substring(0, dot) : version;
             eeVersion = Integer.parseInt(major);
         }
+        eeVersionRef = ref;
     }
 
     /**
@@ -141,6 +144,9 @@ public class HealthExecutorImpl implements HealthExecutor {
      * @param ref reference to the service
      */
     protected void unsetEEVersion(ServiceReference<JavaEEVersion> ref) {
-        eeVersion = 0;
+        if (eeVersionRef == ref) {
+            eeVersionRef = null;
+            eeVersion = 0;
+        }
     }
 }

--- a/dev/com.ibm.ws.transaction.context/bnd.bnd
+++ b/dev/com.ibm.ws.transaction.context/bnd.bnd
@@ -26,11 +26,12 @@ Service-Component: \
     implementation:=com.ibm.ws.transaction.context.internal.TransactionContextProviderImpl; \
     provide:='com.ibm.wsspi.threadcontext.ThreadContextProvider,com.ibm.wsspi.threadcontext.jca.JCAContextProvider'; \
     configuration-policy:=ignore;\
-    EEVersion='com.ibm.ws.javaee.version.JavaEEVersion';\
+    eEVersion='com.ibm.ws.javaee.version.JavaEEVersion?';\
     transactionInflowManager=com.ibm.tx.jta.TransactionInflowManager;\
     transactionManager=com.ibm.ws.tx.embeddable.EmbeddableWebSphereTransactionManager; \
-    greedy:='EEVersion';\
-    optional:='EEVersion';\
+    dynamic:='eEVersion';\
+    greedy:='eEVersion';\
+    optional:='eEVersion';\
     properties:='\
      alwaysCaptureThreadContext:Boolean=true,\
      type=javax.resource.spi.work.TransactionContext'

--- a/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextProviderImpl.java
+++ b/dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/TransactionContextProviderImpl.java
@@ -35,9 +35,14 @@ import com.ibm.wsspi.threadcontext.jca.JCAContextProvider;
  */
 public class TransactionContextProviderImpl implements JCAContextProvider, ThreadContextProvider {
     /**
-     * Jakarta EE versiom if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
+     * Jakarta EE version if Jakarta EE 9 or higher. If 0, assume a lesser EE spec version.
      */
-    int eeVersion;
+    volatile int eeVersion;
+
+    /**
+     * Tracks the most recently bound EE version service reference. Only use this within the set/unsetEEVersion methods.
+     */
+    private ServiceReference<JavaEEVersion> eeVersionRef;
 
     /**
      * Reference to the transaction inflow manager.
@@ -181,6 +186,7 @@ public class TransactionContextProviderImpl implements JCAContextProvider, Threa
             String major = dot > 0 ? version.substring(0, dot) : version;
             eeVersion = Integer.parseInt(major);
         }
+        eeVersionRef = ref;
     }
 
     /**
@@ -205,7 +211,10 @@ public class TransactionContextProviderImpl implements JCAContextProvider, Threa
      * @param ref reference to the service
      */
     protected void unsetEEVersion(ServiceReference<JavaEEVersion> ref) {
-        eeVersion = 0;
+        if (eeVersionRef == ref) {
+            eeVersionRef = null;
+            eeVersion = 0;
+        }
     }
 
     /**

